### ISSUE TITLE
opam install ssl (0.5.5) error

### ISSUE THREAD

**Issue by tempodox** [savonet/ocaml-ssl#34]
opam install ssl (0.5.5) returns an error on Darwin (macOS) that didn't happen with 0.5.4:

```
... CPPFLAGS=-I/usr/local/opt/openssl/include LDFLAGS=-L/usr/local/opt/openssl/lib
# configure: error: Cannot find libssl headers.
```

There is no “/usr/local/opt” in Darwin.

The same error occurs when I create “/usr/local/opt/openssl” and put symbolic links to existing openssl directories there.

Could it be that the “opam” file (new in 0.5.5) is faulty?  The “opam” in the conf-openssl package looks different.

**Comment by toots** [comment]
Ok, I see where this is coming from. Do you also have an error installing `conf-openssl`? You might need to `opam reinstall conf-openssl`. Thanks.

**Comment by tempodox** [comment]
Nope, same error.  `opam reinstall conf-openssl` didn’t change anything, the “opam” file there is exactly the same as before.

I unpinned from ssl-0.5.4 and did `opam upgrade` to get 0.5.5 and have the same error.  Please compare the openssl configs in the “opam” files in the ssl (0.5.5) and conf-openssl packages.  They are different.  My hunch is that the error lies there, but in the ssl package, not in conf-openssl.  ssl-0.5.4 installs flawlessly with the exact same conf-openssl, but then, it doesn’t have its own “opam”.

**Comment by toots** [comment]
My question was wether `conf-openssl` also installed with or without errors. The `opam` file there also reffers to `/usr/opt/local`:
```
build: [
  ["pkg-config" "openssl"] {os != "darwin" & os != "freebsd" & os != "openbsd" & os != "netbsd"}
  ["sh" "-exec" "PKG_CONFIG_PATH=/usr/local/opt/openssl/lib/pkgconfig pkg-config openssl"] {os = "darwin"}
]
```
I wrote a detailed explanation of the situation in ocaml/opam-repository#10453

Let's start by the beginning then:
* Which version of `macOS` are you using?
* Did you install `openssl` through `homebrew`?
* What do you get when you execute: `brew info openssl`

Thanks!

**Comment by tempodox** [comment]
Yes, conf-openssl has always installed without error.

The macOS version (10.12.6) seems to be irrelevant, since I can install ssl-0.5.4 right now without error by version-pinning it, on my same OS, with the same conf-openssl.

openssl is (and always was) installed via MacPorts, and it seems to lack a pkg-config (doing `pkg-config openssl` outputs nothing).

The one conspicuous difference between ssl-0.5.4 and ssl-0.5.5 is that the latter has an “opam” file where the former didn’t.  And the openssl config there differs from conf-openssl. So I thought that might be a viable lead, since the build fails in the “configure” stage.

**Comment by toots** [comment]
The local `opam` file in `ocaml-ssl` is not being used. Your issues comes from the fact that `openssl` support in `opam`-packaged `ssl` is based on the `homebrew` version of `openssl`.

Previous versions were compiling/installing because of a bug in `ocaml-ssl` where it did not honor the given `LDFLAGS` config, resulting in improper linking against the wrong library.

`pkg-config ssl` is supposed to return nothing. It exits with error code `1` when `ssl` cannot be found. Here:
```
% pkg-config openssl
zsh: exit 1     pkg-config openssl
% PKG_CONFIG_PATH=/usr/local/opt/openssl/lib/pkgconfig pkg-config openssl
(nothing)
```

I would like to properly address this but I'm gonna need a couple of input from you:
* Are you sure that you are actually reinstalling the latest `conf-openssl`? This package was changed as recently as yesterday I think and I would hate to miss an issue there while fixing your issue. Try:
```
opam update
opam reinstall --verbose conf-openssl
```
Latest version should log:
```
[conf-openssl: sh] Command started
+ sh "-exec" "PKG_CONFIG_PATH=/usr/local/opt/openssl/lib/pkgconfig pkg-config openssl" (CWD=/Volumes/Home/toots/.opam/4.04.0+32bit/build/conf-openssl.1)
- + PKG_CONFIG_PATH=/usr/local/opt/openssl/lib/pkgconfig
- + pkg-config openssl
∗  installed conf-openssl.1
```

* Do you know the content of the MacPorts `openssl` package? Where does it ship `openssl` headers and shared objects?

Thanks!

**Comment by tempodox** [comment]
I have conf-openssl.1 installed, but I did the `upgrade` & `reinstall --verbose` nevertheless just to be sure.  It says what you show in your mail and the error persists.

MacPort’s openssl paths are,

headers:
/opt/local/include/openssl

shared objects:
/opt/local/lib

(it has “libssl.1.0.0.dylib” and “libssl.a”)


I looked and there is actually an “openssl.pc” in “/opt/local/lib/pkgconfig”, and the exit code of `pkg-config openssl` is 0 on my system, while outputting nothing.

**Comment by toots** [comment]
Excellent, thanks. I'll fix this ASAP.

**Comment by toots** [comment]
Ok, I have pushed a fix. Could you try the following:
```
git clone --recursive https://github.com/savonet/ocaml-ssl.git
cd ocaml-ssl
opam pin add ssl .
```
You will need `autoconf` and `automake`. Looks like macport has ports for each of those.

If you can confirm with me that this works, I'll push the updates to the main `opam` repository.

**Comment by tempodox** [comment]
I had to tweak it a bit, but your fix is working.  In the freshly cloned “ocaml-ssl” dir, I did the following (“$” is the shell prompt):

$ ./bootstrap

$ ./configure --includedir='/opt/local/include' --libdir='/opt/local/lib' CPPFLAGS='-I/opt/local/include' LDFLAGS='-L/opt/local/lib’

$ make
# success! #

For the `opam pin` to succeed, I had to edit “./opam/files/osx-configure.sh” a bit.  The “-I” option in the CPPFLAGS variable must be “/opt/local/include” for the MacPorts branch, since C code inserts the “openssl” subdirectory in its `include` directive.  So, the same “-I” as for `configure` in “osx-configure.sh”.

Thank you very much :)
